File: src/ffi/grovedb/src/get.cpp
// Copyright (c) 2026-present, The Dash Core developers
// Distributed under the MIT software license, see the accompanying
// file LICENSE.MIT or https://opensource.org/license/mit

#include <types/db.h>
#include <types/transaction.h>
#include <util/ffi.h>

#include <grovedb/wire.h>

namespace grovedb {
// ---------------------------------------------------------------------------
// Get — follows references
// ---------------------------------------------------------------------------

Result<Costed<Element>, Error> Db::Get(const Path& path, const Bytes& key)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<Element>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_get(**m_impl->m_db, ToSlice(path_buf), ToSlice(key));
    return ConvertElement(result);
  });
}

Result<Costed<Element>, Error> Db::Get(const Path& path, const Bytes& key, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<Element>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_get_with_tx(
        **m_impl->m_db, ToSlice(path_buf), ToSlice(key), *txn.m_impl->m_tx
    );
    return ConvertElement(result);
  });
}

// ---------------------------------------------------------------------------
// GetDirect — no reference following
// ---------------------------------------------------------------------------

Result<Costed<Element>, Error> Db::GetDirect(const Path& path, const Bytes& key)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<Element>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_get_raw(**m_impl->m_db, ToSlice(path_buf), ToSlice(key));
    return ConvertElement(result);
  });
}

Result<Costed<Element>, Error>
Db::GetDirect(const Path& path, const Bytes& key, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<Element>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_get_raw_with_tx(
        **m_impl->m_db, ToSlice(path_buf), ToSlice(key), *txn.m_impl->m_tx
    );
    return ConvertElement(result);
  });
}

// ---------------------------------------------------------------------------
// GetOptional — returns nullopt when missing
// ---------------------------------------------------------------------------

Result<Costed<std::optional<Element>>, Error> Db::GetOptional(const Path& path, const Bytes& key)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<std::optional<Element>>, Error> {
    auto path_buf = wire::Encode(path);
    auto result =
        grovedb_cxx::grovedb_get_raw_optional(**m_impl->m_db, ToSlice(path_buf), ToSlice(key));
    return ConvertOptionalElement(result);
  });
}

Result<Costed<std::optional<Element>>, Error>
Db::GetOptional(const Path& path, const Bytes& key, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<std::optional<Element>>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_get_raw_optional_with_tx(
        **m_impl->m_db, ToSlice(path_buf), ToSlice(key), *txn.m_impl->m_tx
    );
    return ConvertOptionalElement(result);
  });
}

// ---------------------------------------------------------------------------
// KeyExists — existence check
// ---------------------------------------------------------------------------

Result<Costed<bool>, Error> Db::KeyExists(const Path& path, const Bytes& key)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_has_raw(**m_impl->m_db, ToSlice(path_buf), ToSlice(key));
    return ConvertBool(result);
  });
}

Result<Costed<bool>, Error>
Db::KeyExists(const Path& path, const Bytes& key, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_has_raw_with_tx(
        **m_impl->m_db, ToSlice(path_buf), ToSlice(key), *txn.m_impl->m_tx
    );
    return ConvertBool(result);
  });
}

// ---------------------------------------------------------------------------
// SubtreeExists — subtree/path validation
// ---------------------------------------------------------------------------

Result<Costed<bool>, Error> Db::SubtreeExists(const Path& path)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_check_subtree_exists(**m_impl->m_db, ToSlice(path_buf));
    return ConvertBool(result);
  });
}

Result<Costed<bool>, Error> Db::SubtreeExists(const Path& path, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_check_subtree_exists_with_tx(
        **m_impl->m_db, ToSlice(path_buf), *txn.m_impl->m_tx
    );
    return ConvertBool(result);
  });
}

// ---------------------------------------------------------------------------
// IsEmptyTree
// ---------------------------------------------------------------------------

Result<Costed<bool>, Error> Db::IsEmptyTree(const Path& path)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_is_empty_tree(**m_impl->m_db, ToSlice(path_buf));
    return ConvertBool(result);
  });
}

Result<Costed<bool>, Error> Db::IsEmptyTree(const Path& path, const Transaction& txn)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<Costed<bool>, Error> {
    auto path_buf = wire::Encode(path);
    auto result = grovedb_cxx::grovedb_is_empty_tree_with_tx(
        **m_impl->m_db, ToSlice(path_buf), *txn.m_impl->m_tx
    );
    return ConvertBool(result);
  });
}
} // namespace grovedb

File: rust/grovedb_cxx/src/get.rs
//
// Copyright (c) 2026-present, The Dash Core developers
// SPDX-License-Identifier: MIT
// See the accompanying file LICENSE or https://opensource.org/license/mit
//

use grovedb_path::SubtreePath;
use grovedb_version::version::GroveVersion;

use crate::element::serialize_element;
use crate::ffi::{FfiBoolResult, FfiElementResult, FfiOptionalElementResult};
use crate::lifecycle::operation_cost_to_ffi;
use crate::types::decode_path;
use crate::BoxedGroveDb;
use crate::BoxedTransaction;

// ---------------------------------------------------------------------------
// get — follows references
// ---------------------------------------------------------------------------

/// Get an element by path and key, following any references to their target.
pub fn grovedb_get(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
) -> Result<FfiElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db.db.get(segments.as_slice(), key, None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let element = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiElementResult {
    element: serialize_element(&element, version)?,
    cost,
  })
}

/// Get an element by path and key within a transaction, following references.
pub fn grovedb_get_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .get(segments.as_slice(), key, Some(&tx.tx), version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let element = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiElementResult {
    element: serialize_element(&element, version)?,
    cost,
  })
}

// ---------------------------------------------------------------------------
// get_raw — no reference following
// ---------------------------------------------------------------------------

/// Get an element by path and key without following references.
pub fn grovedb_get_raw(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
) -> Result<FfiElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .get_raw(segments.as_slice().into(), key, None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let element = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiElementResult {
    element: serialize_element(&element, version)?,
    cost,
  })
}

/// Get an element by path and key within a transaction, without following references.
pub fn grovedb_get_raw_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .get_raw(segments.as_slice().into(), key, Some(&tx.tx), version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let element = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiElementResult {
    element: serialize_element(&element, version)?,
    cost,
  })
}

// ---------------------------------------------------------------------------
// get_raw_optional — returns None instead of erroring when missing
// ---------------------------------------------------------------------------

/// Get an element optionally: returns `has_element = false` instead of an
/// error when the key is missing.
pub fn grovedb_get_raw_optional(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
) -> Result<FfiOptionalElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .get_raw_optional(segments.as_slice().into(), key, None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let maybe = ctx.value.map_err(crate::ffi_error)?;
  match maybe {
    Some(element) => Ok(FfiOptionalElementResult {
      has_element: true,
      element: serialize_element(&element, version)?,
      cost,
    }),
    None => Ok(FfiOptionalElementResult {
      has_element: false,
      element: Vec::new(),
      cost,
    }),
  }
}

/// Get an element optionally within a transaction.
pub fn grovedb_get_raw_optional_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiOptionalElementResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .get_raw_optional(segments.as_slice().into(), key, Some(&tx.tx), version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let maybe = ctx.value.map_err(crate::ffi_error)?;
  match maybe {
    Some(element) => Ok(FfiOptionalElementResult {
      has_element: true,
      element: serialize_element(&element, version)?,
      cost,
    }),
    None => Ok(FfiOptionalElementResult {
      has_element: false,
      element: Vec::new(),
      cost,
    }),
  }
}

// ---------------------------------------------------------------------------
// has_raw — existence check
// ---------------------------------------------------------------------------

/// Check whether a key exists at the given path (no reference following).
pub fn grovedb_has_raw(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db.db.has_raw(segments.as_slice(), key, None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let exists = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiBoolResult {
    value: exists,
    cost,
  })
}

/// Check whether a key exists at the given path within a transaction.
pub fn grovedb_has_raw_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  key: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .has_raw(segments.as_slice(), key, Some(&tx.tx), version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let exists = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiBoolResult {
    value: exists,
    cost,
  })
}

// ---------------------------------------------------------------------------
// check_subtree_exists — subtree/path validation
// ---------------------------------------------------------------------------

/// Check whether all parent subtrees in the path exist.
///
/// Returns `value = true` when the path is valid, `value = false` when any
/// parent subtree is missing.  Only returns `Err` for path-encoding errors.
pub fn grovedb_check_subtree_exists(
  db: &BoxedGroveDb,
  path: &[u8],
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db
    .db
    .check_subtree_exists_invalid_path(segments.as_slice().into(), None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  Ok(FfiBoolResult {
    value: ctx.value.is_ok(),
    cost,
  })
}

/// Check whether all parent subtrees in the path exist, within a transaction.
pub fn grovedb_check_subtree_exists_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let ctx = db.db.check_subtree_exists_invalid_path(
    segments.as_slice().into(),
    Some(&tx.tx),
    version,
  );
  let cost = operation_cost_to_ffi(&ctx.cost);
  Ok(FfiBoolResult {
    value: ctx.value.is_ok(),
    cost,
  })
}

// ---------------------------------------------------------------------------
// is_empty_tree — check whether a subtree is empty
// ---------------------------------------------------------------------------

/// Check whether the subtree at the given path is empty.
pub fn grovedb_is_empty_tree(
  db: &BoxedGroveDb,
  path: &[u8],
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let subtree_path: SubtreePath<Vec<u8>> = segments.as_slice().into();
  let ctx = db.db.is_empty_tree(subtree_path, None, version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let empty = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiBoolResult {
    value: empty,
    cost,
  })
}

/// Check whether the subtree at the given path is empty, within a transaction.
pub fn grovedb_is_empty_tree_with_tx(
  db: &BoxedGroveDb,
  path: &[u8],
  tx: &BoxedTransaction,
) -> Result<FfiBoolResult, String> {
  let version = GroveVersion::latest();
  let segments = decode_path(path)?;
  let subtree_path: SubtreePath<Vec<u8>> = segments.as_slice().into();
  let ctx = db.db.is_empty_tree(subtree_path, Some(&tx.tx), version);
  let cost = operation_cost_to_ffi(&ctx.cost);
  let empty = ctx.value.map_err(crate::ffi_error)?;
  Ok(FfiBoolResult {
    value: empty,
    cost,
  })
}
